tensorflow/python/keras/layers/normalization.py

```diff
@@ -613,8 +613,50 @@ def _fake_update():
 
     return (r, d, out_mean, out_variance)
 
-  def _moments(self, inputs, reduction_axes, keep_dims):
-    mean, variance = nn.moments(inputs, reduction_axes, keep_dims=keep_dims)
+  def moment_calculation(self,
+                         inputs,
+                         axes,
+                         keep_dims=False,
+                         mask=None):
+
+      with ops.name_scope("moments", values=[inputs, axes]):
+          # The dynamic range of fp16 is too limited to support the collection of
+          # sufficient statistics. As a workaround we simply perform the operations
+          # on 32-bit floats before converting the mean and variance back to fp16
+          casted_inputs = math_ops.cast(inputs, dtypes.float32) if inputs.dtype == dtypes.float16 else inputs
+          # Compute true mean while keeping the dims for proper broadcasting.
+
+          if mask is not None:
+              mask = array_ops.expand_dims(mask, axis=-1)
+              binary_mask = math_ops.cast(mask, casted_inputs.dtype)
```

Comment:
After the cast, it is not binary anymore, so probably rename it. Also, the mask is never used below, only the binary_mask is used. How about just:

mask = math_ops.cast(array_ops.expand_dims(mask, axis=-1), casted_inputs.dtype)